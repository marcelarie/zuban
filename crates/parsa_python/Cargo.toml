[package]
name = "parsa_python"
version.workspace = true
edition.workspace = true
rust-version.workspace = true
publish = false
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
test_utils.workspace = true
