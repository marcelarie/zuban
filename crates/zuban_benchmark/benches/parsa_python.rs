use criterion::{criterion_group, criterion_main, Criterion};

use parsa_python_cst::Tree;

fn parse_file(copies: usize) {
    let some_code = utils::dedent(
        r#"
        import os
        def x(a, b, c):
            try:
                return a
            except Exception:
                return [1, 2, 3, 4, 5, 6]
        class C:
            x: int
            y = 1 # type: int
            z: dict[int, list[str]] = {1: ["", r''], 2: [f""" {1} """]}
            def f(self, x: list[str], y: int): ...
    "#,
    );
    assert_eq!(some_code.len(), 262);
    Tree::parse(some_code.repeat(copies).into());
}

fn bench_parser(c: &mut Criterion) {
    c.bench_function("parse big file", |b| b.iter(|| parse_file(1000)));
    c.bench_function("parse small file", |b| b.iter(|| parse_file(1)));
}

// Register the benchmarks
criterion_group!(benches, bench_parser);
criterion_main!(benches);
