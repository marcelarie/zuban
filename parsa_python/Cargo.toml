[package]
name = "parsa_python"
version = "0.1.0"
edition.workspace = true
rust-version.workspace = true
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
