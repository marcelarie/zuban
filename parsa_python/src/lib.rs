use parsa::{create_parser,create_node,create_token};


create_token!(PythonToken, enum TokenType, String, Number, Endmarker);
create_node!(PythonNode, enum NodeType, file_input, foo);

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn it_works() {
        assert_eq!(2 + 2, 4);
        parse_python("bar");
        dbg!(TokenType::get_map());
        dbg!(NodeType::get_map());
        return
    }
}

struct PythonTokenizer {
}

impl PythonTokenizer {
    fn new_tok(&self) -> PythonToken {
        PythonToken {
            start: 0,
            length: 0,
            type_: TokenType::Endmarker,
        }
    }
}

impl parsa::Tokenizer<PythonToken> for PythonTokenizer {
    fn new(_code: &str) -> Self {
        Self {}
    }

    fn yield_next(&self) -> PythonToken {
        self.new_tok()
    }

}

create_parser!(parse_python, PythonTree, PythonNode,
               PythonToken, PythonTokenizer, TokenType, NodeType);
//create_parser!(parse_pythonx, PythonTreex, PythonNodex,
//               PythonToken, PythonTokenizer, TokenType, NodeType);
