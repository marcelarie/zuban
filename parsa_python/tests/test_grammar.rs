extern crate insta;

use parsa_python::*;

fn dedent(s: &'static str) -> String {
    let mut indent = usize::MAX;
    let lines: &Vec<_> = &s.split('\n').collect();
    for line in lines {
        if !line.trim().is_empty() {
            indent = std::cmp::min(indent, line.len() - line.trim_start().len());
        }
    }
    if indent == usize::MAX {
        return s.to_string();
    }
    let new_lines: Vec<_> = lines
        .iter()
        .map(|line| {
            if line.len() >= indent {
                &line[indent..]
            } else {
                line
            }
        })
        .collect();
    new_lines.join("\n")
}

fn tree_to_string(tree: PyTree) -> String {
    fn recurse(code: &mut String, node: &PyNode, depth: usize) {
        *code += &" ".repeat(depth);
        *code += &format!(
            "{}: {}-{}{}\n",
            node.type_str(),
            node.start(),
            node.end(),
            if node.is_leaf() {
                format!(" {:?}", node.as_code())
            } else {
                "".to_string()
            }
        );
        for c in node.iter_children() {
            assert_eq!(node.index, c.parent().unwrap().index);
            recurse(code, &c, depth + 1);
        }
    }

    let root = tree.root_node();
    assert!(root.parent().is_none());
    let mut code = String::new();
    recurse(&mut code, &root, 0);
    code
}

macro_rules! parametrize_snapshots {
    ($($name:ident : $input:expr;)*) => {$(
        #[test]
        fn $name() {
            let tree = parse($input.into());
            insta::assert_snapshot!(stringify!($name), tree_to_string(tree));
        }
    )*}
}

parametrize_snapshots!(
    empty: "".to_owned();
    very_simple: "1".to_owned();
    json: "{foo: 1}\n".to_owned();
    simple: dedent("
        a(bar:=1, foo=1);
        if a:
         1
        else: 2
        ");
    assignments_simple: dedent("
        a = 1
        a, b = 2
        a, b, = 2
        (a, b) = 2
        ((a, b)) = 2
        a = b = 3
        a = (b, c) = 3
        a = (b, c,) = 3
        a = b, c, = 3
        ");
    assignments_star: dedent("
        *foo = *bar = *baz
        i, *foo = ((j, [*bar, k])) = *baz
        *foo, *bar = yield baz
        ");
    assignments_aug: dedent("
        foo += 3
        foo.bar |= 3
        foo, bar += 3
        ");
    assignments_setitem: dedent("
        foo[0] += 3
        foo[0] = a = 3
        foo[a, b] = c
        foo[a::, b:f, c:=3] = c
        ");
    assignments_annotation: dedent("
        foo: bar
        foo: int = 3
        foo.bar: baz
        foo: bar = a = 4
        ");
    assignments_with_call: dedent("
        foo().bar = 1
        foo(1,).bar[1] = 1
        foo(a for a in b).bar[1] = 1
        ");
    for_stmt: dedent("
        for x in [1,2]:
            ...
        for *b in *c: ...
        for c.x, (*b, d.y[0]) in i, *j, k: ...
        ");
    comprehension_simple: dedent("
        [1 for a in b]
        [a for *a, c in b if c if d]
        foo(a for (*a, (*c, b),), in d)
        ");
    comprehension_fails: dedent("
        # fails
        [1 for a in *b]
        [1 for a in b if a else c]
        ");
    del_stmt: dedent("
        del foo
        del foo,
        del foo, bar.baz
        del foo[1]
        ");
    cls: dedent("
        class Foo(object):
            def __init__(self, /, f, *, g):
                pass

            @property
            def prop(self):
                pass
        ");
    import_names: dedent("
        import foo
        import foo.bar as baz
        import foo, bar, baz
        import foo.bar.baz
        ");
    import_name_fails: dedent("
        import foo,
        import foo, bar,
        ");
    import_from_normal: dedent("
        from foo import bar
        from foo import bar as baz
        from foo.bar.baz import *
        ");
    import_from_dotted: dedent("
        from . import (foo as baz, bar)
        from . import (foo, bar,)
        from ..foo import bar as baz
        from ... import *
        ");
    import_from_fails: dedent("
        from . import foo,
        from . import foo, bar,
        ");
    calls: dedent("
        foo(bar)
        foo(bar_foo=3)
        foo(bar_baz:=2)
        foo(**kwargs)
        ");
    nested_call: dedent("
        foo(1)()
        ");
    complex_calls: dedent("
        foo(bar, 3)
        foo(bar, *foo)
        foo(bar, *foo, name=3)
        foo(bar, *foo, name=3, **kwargs)
        foo(bar, *foo, name=3, **kwargs, b=3)
        foo(**kwargs, b=3, **foo)
        ");
    operators: dedent("
        f | d | c + 5 & 2 * 3 * 4 ** 5 ** 6
        ");
    simple_error_recovery: dedent("
        a + 3 +
        b = 3
        ");
    terminal_error_recovery: dedent("
        ?
        ");
    terminal_and_nonterminal_error_recovery: dedent("
        foo
        (1) + ? + c
        bar
        ");
    keyword_recovery: dedent("
        foo
        def
        bar
        else:
        baz
        ");
    match_simple: dedent("
        match foo:
            case ['a']: ...
            case _:
                pass
        ");
    soft_keyword_underscore: dedent("
        match _:
            case ['a']: _(_)
            case _:
                _
        match = 3
        ");
    match_underscore_with_as: dedent("
        match foo:
            case bar as _: pass
        ");
    match_class_pattern: dedent("
        match a:
            case A():
                pass
            case B(1, 2):
                pass
            case B(1, b=2):
                pass
            case B(a=1, b=2):
                pass
        ");
    dict_literal1: dedent("
        {1: 2}
        {**foo}
        {**foo, 1: 2}
        {**foo, **bar}
        ");
    function_simple: dedent("
        def foo(): ...
        def foo(a: 3): ...
        def foo(a, b,): ...
        def foo(a: f = 3, b=4): ...
        def foo(a=3, b=4): ...
        ");
    function_args: dedent("
        def foo(a, *, b=3): ...
        def foo(a=3, *, c): ...
        def foo(a, *, x, **kwargs): ...
        def foo(a, *args: int, **fooo: str,): ...
        def foo(a, /, b: i, c: i=3, **kwargs): ...
        def foo(*args): ...
        ");
    lambda_simple: dedent("
        lambda: ...
        lambda a: ...
        lambda a, b,: ...
        lambda a=3, b=4: ...
        ");
    lambda_slash: dedent("
        lambda a, /, b=3: ...
        lambda a=3, /,: ...
        lambda a=3, /: ...
        lambda a=3, /, *args, c=3: ...
        lambda a=3, /, c=3: ...
        lambda a, /, b, c=3, **kwargs: ...
        lambda a, a=3, /, c=3, *, d: ...
        ");
    lambda_args: dedent("
        lambda a, *, b=3: ...
        lambda a=3, *, c: ...
        lambda a, *, x, **kwargs: ...
        lambda a, *args, **fooo ,: ...
        lambda *args: ...
        ");
    lambda_kwargs: dedent("
        lambda **kwargs: ...
        lambda a=3, **kwargs: ...
        lambda a, **kwargs: ...
        lambda a, /, b, *, x, **kwargs: ...
        ");
    lambda_fails: dedent("
        lambda a, *, **kwargs: ...
        lambda a=3, b: ...
        lambda **a, b: ...
        lambda a=3, *,: ...
        lambda ,: ...
        ");
    lambda_fail_with_slash: dedent("
        lambda a=3, /, b: ...
        lambda a=3, /, b=3, c: ...
        lambda a, /, b=3, c: ...
        lambda z, a=3, d, /, b: ...
        ");
    with_items_simple: dedent("
        with foo as bar: ...
        with foo as bar, baz: ...
        ");
    with_items_parentheses: dedent("
        with (foo as bar): ...
        with (foo as bar, baz): ...
        with ((foo)): ...
        with (i for i in 42): ...
        with (yield y): ...
        ");
    string_literals_in_strings: dedent(r#"
        '"""'
        "'''"
        ""
        ''
        "#);
    bytes: dedent(r#"
        b'asdf\n\''
        br'\asdf'
        Br'asdf"""'
        bR''
        BR'\''
        "#);
    exception_group: dedent(r#"
        try:
            ...
        except* SpamError:
            ...
        except* FooError as e:
            ...
        except* (BarError, BazError) as e:
            ...
        "#);
    invalid_syntax: dedent(r#"
        def f():
            (x,) += 1
            z
        "#);
);
