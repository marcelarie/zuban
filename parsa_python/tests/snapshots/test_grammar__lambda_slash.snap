---
source: parsa_python/tests/test_grammar.rs
expression: tree_to_string(tree)
---
file: 0-185
 stmt: 1-23
  simple_stmts: 1-23
   simple_stmt: 1-22
    star_expressions: 1-22
     expression: 1-22
      lambda: 1-22
       Keyword: 1-7 "lambda"
       lambda_parameters: 8-17
        lambda_param_no_default: 8-9
         name_definition: 8-9
          Name: 8-9 "a"
        Keyword: 9-10 ","
        Keyword: 11-12 "/"
        Keyword: 12-13 ","
        lambda_param_with_default: 14-17
         name_definition: 14-15
          Name: 14-15 "b"
         Keyword: 15-16 "="
         expression: 16-17
          bitwise_or: 16-17
           Number: 16-17 "3"
       Keyword: 17-18 ":"
       expression: 19-22
        bitwise_or: 19-22
         Keyword: 19-22 "..."
   Newline: 22-23 "\n"
 stmt: 23-43
  simple_stmts: 23-43
   simple_stmt: 23-42
    star_expressions: 23-42
     expression: 23-42
      lambda: 23-42
       Keyword: 23-29 "lambda"
       lambda_parameters: 30-37
        lambda_param_with_default: 30-33
         name_definition: 30-31
          Name: 30-31 "a"
         Keyword: 31-32 "="
         expression: 32-33
          bitwise_or: 32-33
           Number: 32-33 "3"
        Keyword: 33-34 ","
        Keyword: 35-36 "/"
        Keyword: 36-37 ","
       Keyword: 37-38 ":"
       expression: 39-42
        bitwise_or: 39-42
         Keyword: 39-42 "..."
   Newline: 42-43 "\n"
 stmt: 43-62
  simple_stmts: 43-62
   simple_stmt: 43-61
    star_expressions: 43-61
     expression: 43-61
      lambda: 43-61
       Keyword: 43-49 "lambda"
       lambda_parameters: 50-56
        lambda_param_with_default: 50-53
         name_definition: 50-51
          Name: 50-51 "a"
         Keyword: 51-52 "="
         expression: 52-53
          bitwise_or: 52-53
           Number: 52-53 "3"
        Keyword: 53-54 ","
        Keyword: 55-56 "/"
       Keyword: 56-57 ":"
       expression: 58-61
        bitwise_or: 58-61
         Keyword: 58-61 "..."
   Newline: 61-62 "\n"
 stmt: 62-93
  simple_stmts: 62-93
   simple_stmt: 62-92
    star_expressions: 62-92
     expression: 62-92
      lambda: 62-92
       Keyword: 62-68 "lambda"
       lambda_parameters: 69-87
        lambda_param_with_default: 69-72
         name_definition: 69-70
          Name: 69-70 "a"
         Keyword: 70-71 "="
         expression: 71-72
          bitwise_or: 71-72
           Number: 71-72 "3"
        Keyword: 72-73 ","
        Keyword: 74-75 "/"
        Keyword: 75-76 ","
        lambda_star_etc: 77-87
         Keyword: 77-78 "*"
         lambda_starred_param: 78-82
          name_definition: 78-82
           Name: 78-82 "args"
         Keyword: 82-83 ","
         lambda_param_maybe_default: 84-87
          name_definition: 84-85
           Name: 84-85 "c"
          Keyword: 85-86 "="
          expression: 86-87
           bitwise_or: 86-87
            Number: 86-87 "3"
       Keyword: 87-88 ":"
       expression: 89-92
        bitwise_or: 89-92
         Keyword: 89-92 "..."
   Newline: 92-93 "\n"
 stmt: 93-117
  simple_stmts: 93-117
   simple_stmt: 93-116
    star_expressions: 93-116
     expression: 93-116
      lambda: 93-116
       Keyword: 93-99 "lambda"
       lambda_parameters: 100-111
        lambda_param_with_default: 100-103
         name_definition: 100-101
          Name: 100-101 "a"
         Keyword: 101-102 "="
         expression: 102-103
          bitwise_or: 102-103
           Number: 102-103 "3"
        Keyword: 103-104 ","
        Keyword: 105-106 "/"
        Keyword: 106-107 ","
        lambda_param_with_default: 108-111
         name_definition: 108-109
          Name: 108-109 "c"
         Keyword: 109-110 "="
         expression: 110-111
          bitwise_or: 110-111
           Number: 110-111 "3"
       Keyword: 111-112 ":"
       expression: 113-116
        bitwise_or: 113-116
         Keyword: 113-116 "..."
   Newline: 116-117 "\n"
 stmt: 117-152
  simple_stmts: 117-152
   simple_stmt: 117-151
    star_expressions: 117-151
     expression: 117-151
      lambda: 117-151
       Keyword: 117-123 "lambda"
       lambda_parameters: 124-146
        lambda_param_no_default: 124-125
         name_definition: 124-125
          Name: 124-125 "a"
        Keyword: 125-126 ","
        Keyword: 127-128 "/"
        Keyword: 128-129 ","
        lambda_param_no_default: 130-131
         name_definition: 130-131
          Name: 130-131 "b"
        Keyword: 131-132 ","
        lambda_param_with_default: 133-136
         name_definition: 133-134
          Name: 133-134 "c"
         Keyword: 134-135 "="
         expression: 135-136
          bitwise_or: 135-136
           Number: 135-136 "3"
        Keyword: 136-137 ","
        lambda_star_etc: 138-146
         lambda_double_starred_param: 138-146
          Keyword: 138-140 "**"
          name_definition: 140-146
           Name: 140-146 "kwargs"
       Keyword: 146-147 ":"
       expression: 148-151
        bitwise_or: 148-151
         Keyword: 148-151 "..."
   Newline: 151-152 "\n"
 stmt: 152-185
  simple_stmts: 152-185
   simple_stmt: 152-184
    star_expressions: 152-184
     expression: 152-184
      lambda: 152-184
       Keyword: 152-158 "lambda"
       lambda_parameters: 159-179
        lambda_param_no_default: 159-160
         name_definition: 159-160
          Name: 159-160 "a"
        Keyword: 160-161 ","
        lambda_param_with_default: 162-165
         name_definition: 162-163
          Name: 162-163 "a"
         Keyword: 163-164 "="
         expression: 164-165
          bitwise_or: 164-165
           Number: 164-165 "3"
        Keyword: 165-166 ","
        Keyword: 167-168 "/"
        Keyword: 168-169 ","
        lambda_param_with_default: 170-173
         name_definition: 170-171
          Name: 170-171 "c"
         Keyword: 171-172 "="
         expression: 172-173
          bitwise_or: 172-173
           Number: 172-173 "3"
        Keyword: 173-174 ","
        lambda_star_etc: 175-179
         Keyword: 175-176 "*"
         Keyword: 176-177 ","
         lambda_param_maybe_default: 178-179
          name_definition: 178-179
           Name: 178-179 "d"
       Keyword: 179-180 ":"
       expression: 181-184
        bitwise_or: 181-184
         Keyword: 181-184 "..."
   Newline: 184-185 "\n"
 Endmarker: 185-185 ""

